core/trino-spi/src/main/java/io/trino/spi/predicate/Utils.java

```diff
@@ -64,4 +72,56 @@ static RuntimeException handleThrowable(Throwable throwable)
         }
         return new RuntimeException(throwable);
     }
+
+    /**
+     * Try to expand {@code valueSet} into a discrete set of (at most {@code limit}) objects.
+     * For example: [1, 5] can be expanded into {1, 2, 3, 4, 5}.
+     * If the data type is not supported or the expansion results in too many values, {@code Optional.empty()} is returned.
+     */
+    public static Optional<Collection<Object>> tryExpandRanges(ValueSet valuesSet, int valuesLimit)
+    {
+        if (valuesLimit < 0) {
+            throw new IllegalArgumentException("valuesLimit cannot be negative");
+        }
+        if (!(valuesSet instanceof SortedRangeSet) || valuesLimit == 0) {
+            return Optional.empty();
+        }
+        List<Range> ranges = valuesSet.getRanges().getOrderedRanges();
+        Type type = valuesSet.getType();
+        List<Object> result = new ArrayList<>(valuesLimit);
+        for (Range range : ranges) {
+            if (range.isSingleValue()) {
+                result.add(range.getSingleValue());
+                if (result.size() > valuesLimit) {
```

Comment:
check before adding. Since you need to check anyway, it saves you one element.... and an array resize.